src/lib/Type.hs

```diff
@@ -115,9 +115,29 @@ getClassTy (ClassDef _ _ bs _ _) = go bs
 
 -- === querying effects ===
 
-exprEffects :: (MonadFail1 m, EnvReader m) => Expr n -> m n (EffectRow n)
+computeAbsEffects :: (EnvExtender m, SubstE Name e)
+  => Abs (Nest Decl) e n -> m n (Abs (Nest Decl) (EffectRow `PairE` e) n)
+computeAbsEffects it = refreshAbs it \decls result -> do
+  effs <- declNestEffects decls
+  return $ Abs decls (effs `PairE` result)
+
+declNestEffects :: (EnvReader m) => Nest Decl n l -> m l (EffectRow l)
+declNestEffects decls = liftEnvReaderM $ declNestEffectsRec decls
+
+declNestEffectsRec :: Nest Decl n l -> EnvReaderM l (EffectRow l)
+declNestEffectsRec Empty = return mempty
+declNestEffectsRec n@(Nest decl rest) = withExtEvidence n $ do
+  expr <- sinkM $ declExpr decl
+  deff <- exprEffects expr
+  reff <- declNestEffects rest
```

Comment:
The Rec part requires an `EnvExtender` constraint, but when Dougal and I were pair-coding this part, we thought it would be better for the entry point to only require `EnvReader`, since that is semantically more accurate.

But it should recursively call itself (not the non-recursive variant), and I also made the tail-recursion change.  Allocations are down ~1% (though that may be confounded with the effects of inlining).